rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,429 @@
+
+After generating a Yubikey keypair, `tsh` will grab the Yubikey's attestation certificate and generate a new attestation statement [slot cert](https://pkg.go.dev/github.com/go-piv/piv-go@v1.9.0/piv#YubiKey.Attest) for the slot in question. These two certificates will be passed to the Auth server during the certificate signing process in order to verify certificates signed with the slot's public key.
+
+In addition to verifying the certificate chain, the attestation will return the private key's PIN and touch policies, which we can check to enforce specific PIN and touch policies across Teleport Clusters and Roles.
+
+#### Attestation enforcement
+
+Attestation will be enforced in a similar way to [Certificate Locking](https://github.com/gravitational/teleport/blob/master/rfd/0009-locking.md). Essentially, a user will not be able to successfully make requests unless the public key on their certificates has been attested, and the attestation properties (pin and touch policies) meet the criteria set by the server (auth preference and role options).
+
+For this flow, we will need to:
+ 1. perform attestation after the user receives their certificates from a successful login
+ 2. store the attestation object in the backend for future checks
+ 3. check if a certificate has a valid attestation in the backend before authorizing any requests
+
+In `tsh login --piv`, attestation will be performed immediately after the signing process with a new gRPC endpoint:
+
+```
+rpc AssertYubikeyPIVSlot(AssertYubikeyPIVSlotRequest) returns (google.protobuf.Empty);
+
+message AssertYubikeySlotRequest {
+    bytes SlotCert = 1;
+    bytes AttestationCert = 2;
+}
+```
+
+Note: if/when we add other PIV device support in the future, we can deprecate `AssertYubikeyPIVSlot` for a more generalized `AssertPIVSlot` request, or add new custom endpoints (like `AssertSolokeyPIVSlot`) if it cannot be easily generalized accross PIV implementations.
+
+`AssertYubikeyPIVSlot` will then check if the attestation passes the current enforcement criteria for the user. If it doesn't an access denied error will be returned describing what criteria were not met. Otherwise, the attestation will be inserted into the backend in the table `/piv_assertions/<public_key>` to verify future requests for corresponding certificates.
+
+#### Auth Service Config or Auth Preference
+
+You can provide a `piv_attestation` field to to a cluster's Auth Preference, in the Auth Service's config file or with a dynamic Cluster Auth Preference object:
+
+```yaml
+auth_service:
+  ...
+  authentication:
+    ...
+    piv_attestation: 
```

Comment:
I am curious, why is this called specifically "piv_attestation" vs just "piv" to enforce PIV login?

Reply:
I think the main reason was that `piv.mode=disabled` would actually just disable attestation, not disable piv login. I'll just change this to `piv` and `attestation_mode` instead.